### PR TITLE
SAMZA-2657: Blob Store as backend for Samza State backup and restore

### PR DESCRIPTION
**Pull request by shekhars-li**
This PR contains commits from @prateekm and @shekhars-li 

Problem: 
State restore for Samza jobs that have large states takes up a long time, sometimes upto multiple hours because of one message at a time restore for kafka backed changelog. We introduce blob store based backup and restore for Samza, that can do parallel backup and restore and solve the problem discussed earlier.

Solution:
1. Add Blob store under samza-core/storage: Supports the blob store based backup and restore. 
    1. index - Index classes defines a way of accessing a remote file or subdirectory and the metadata associated with it.
         1. FileBlob -> Representation of a blob in blob store with a blob id and offset.
         2. FileIndex -> Representation of a file in blob store as a set of FileBlobs and it's metadata like file name, size, permissions etc. 
        3. DirIndex -> Representation of a directory (and it's sub-directories) in the blob store bucket and metadata like directory name.
        4. SnapshotIndex -> Representation of a snapshot directory and metadata like job name, job id, store name etc.
     2. DirDiff - DirDiff class represents the diff/delta between a local snapshot and a remote snapshot. A corresponding util class DirDiffUtil is used to calculate delta between local and remote snapshots. 

2. Add BlobStoreManager APIs to samza-api/blobstore: 
    1. BlobStoreManager -> Interface to expose GET/PUT/DELETE API calls to a blob store. A special removeTTL API call is introduced to help remove TTL of a blob. Used in garbage collection, as explained in [SAMZA-2657](https://issues.apache.org/jira/browse/SAMZA-2657). 
    2.  Metadata -> Metadata associated with a request to Blob store. Contains job details, store name, payload details. 

Other design related details can be found in the design doc attached with the [SAMZA-2657](https://issues.apache.org/jira/browse/SAMZA-2657)

### CODE REVIEW

File: samza-core/src/main/java/org/apache/samza/storage/blobstore/util/BlobStoreUtil.java [hunk 1]
```diff
@@ -0,0 +1,721 @@
...
+   * identical, as determined by {@link #areSameFile(boolean)}, except for those allowed to differ according to
+   * {@param filesToIgnore}.
+   * 2. The local directory has exactly the same set of sub-directories as the remote directory.
+   *
+   * @param filesToIgnore a set of file names to ignore during the directory comparisons
+   *                      (does not exclude directory names)
+   * @param compareLargeFileChecksums whether to compare checksums for large files (> 1 MB).
+   * @return boolean indicating whether the local and remote directory are identical.
+   */
+  // TODO HIGH shesharm add unit tests
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/storage/blobstore/util/BlobStoreUtil.java:634]: Is this complete? Fix and remove before merging.

File: samza-kv-rocksdb/src/main/java/org/apache/samza/storage/kv/RocksDbOptionsHelper.java [hunk 2]
```diff
@@ -36,9 +40,29 @@
 public class RocksDbOptionsHelper {
   private static final Logger log = LoggerFactory.getLogger(RocksDbOptionsHelper.class);
 
+  public static final String ROCKSDB_WAL_ENABLED = "rocksdb.wal.enabled";
```
Comment (prateekm) [inline samza-kv-rocksdb/src/main/java/org/apache/samza/storage/kv/RocksDbOptionsHelper.java:43]: @shekhars-li Leave a TODO HIGH for me to add these to RocksDBTableDescriptor.

Comment (prateekm) [inline samza-kv-rocksdb/src/main/java/org/apache/samza/storage/kv/RocksDbOptionsHelper.java:43]: @dxichen FYI, there's more here to review below the tests.

File: samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreRestoreManager.java [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+  /**
+   * Starts the restore for the store, enqueuing all restore-completion futures into {@param restoreFutures}.
+   */
+  @VisibleForTesting
+  static void enqueueRestore(String jobName, String jobId, String taskName, String storeName, File storeDir, DirIndex dirIndex,
+      long storeRestoreStartTime, List<CompletionStage<Void>> restoreFutures,
+      BlobStoreUtil blobStoreUtil, BlobStoreRestoreManagerMetrics metrics, ExecutorService executor) {
+    metrics.storePreRestoreNs.get(storeName).set(System.nanoTime() - storeRestoreStartTime);
+
+    Metadata requestMetadata = new Metadata(storeDir.getAbsolutePath(), Optional.empty(), jobName, jobId, taskName, storeName);
```
[the remark this answers is not in the document]
Comment (shekhars-li) [reply samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreRestoreManager.java:315]: Metadata contains Metadata related to the current request. So a restoreDir request has storeDir in payload. When the restoreDir is called, it creates new Metadata object with payload as individual file and that's what is eventually sent to the blob store API (GET). Basically the real request to blob store has file names, because at every step right request Metadata object is built. That way, we track request at each step and use that to log if necessary.

File: samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreStateBackendFactory.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      Clock clock,
+      File loggedStoreBaseDir,
+      File nonLoggedStoreBaseDir) {
+    StorageConfig storageConfig = new StorageConfig(config);
+    String blobStoreManagerFactory = storageConfig.getBlobStoreManagerFactory();
+    Preconditions.checkState(StringUtils.isNotBlank(blobStoreManagerFactory));
+    BlobStoreManagerFactory factory = ReflectionUtil.getObj(blobStoreManagerFactory, BlobStoreManagerFactory.class);
+    BlobStoreManager blobStoreManager = factory.getBackupBlobStoreManager(config, backupExecutor);
+    BlobStoreBackupManagerMetrics metrics = new BlobStoreBackupManagerMetrics(metricsRegistry);
+    BlobStoreUtil blobStoreUtil = new BlobStoreUtil(blobStoreManager, backupExecutor, metrics, null);
```
Comment (dxichen) [inline samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreStateBackendFactory.java:67]: Seems weird to create separate metrics classes for blobStoreUtil since it is agnostic to whether the caller is performing backup or restore, could we merge these metrics classes?

**Review by dxichen — CHANGES_REQUESTED** [review]
Some initial questions, will keep reviewing the rest of the files

File: samza-api/src/main/java/org/apache/samza/storage/blobstore/BlobStoreManager.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * Provides interface for common blob store operations: GET, PUT and DELETE
+ */
+@InterfaceStability.Unstable
+public interface BlobStoreManager {
+  /**
+   * init method to initialize underlying blob store client, if necessary
+   */
+  void init();
```
[the remark this answers is not in the document]
Comment (prateekm) [reply samza-api/src/main/java/org/apache/samza/storage/blobstore/BlobStoreManager.java:37]: Add newline between this and next method too.

File: samza-core/src/main/java/org/apache/samza/config/BlobStoreConfig.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+
+/**
+ * Config related helper methods for BlobStore.
+ */
+public class BlobStoreConfig extends MapConfig {
+  private static final String STORE_PREFIX = "stores.";
+  private static final String CHANGELOG_SUFFIX = ".changelog";
+
+  public static final String DEFAULT_BLOB_STORE_STATE_BACKEND_FACTORY =
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/config/BlobStoreConfig.java:17]: 1. Does this need to be here or can we remove and use class.getName() instead.
2. If it does need to be here, rename field name and remove DEFAULT_. Rename DEFAULT_STORE_STATE_BACKEND_FACTORY to KAFKA_STATE_BACKEND_FACTORY. and change DEFAULT_STORE_BACKUP_FACTORIES to refer to these two properties instead.

File: samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreBackupManager.java [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+    metrics.filesRemaining.getValue().set(0L);
+    metrics.bytesRemaining.getValue().set(0L);
+    metrics.filesToRetain.getValue().set(0L);
+    metrics.bytesToRetain.getValue().set(0L);
+
+    Map<String, CompletableFuture<Pair<String, SnapshotIndex>>>
+        storeToSCMAndSnapshotIndexPairFutures = new HashMap<>();
+    Map<String, CompletableFuture<String>> storeToSerializedSCMFuture = new HashMap<>();
+
+    storesToBackup.forEach((storeName) -> {
```
[the remark this answers is not in the document]
Comment (prateekm) [reply samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreBackupManager.java:168]: This is already async. This is just collecting the futures in the 2 maps above.

File: samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreBackupManager.java [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+        }
+
+        long dirDiffStartTime = System.nanoTime();
+        // get the diff between previous and current store directories
+        DirDiff dirDiff = DirDiffUtil.getDirDiff(checkpointDir, prevDirIndex, BlobStoreUtil.areSameFile(false));
+        metrics.storeDirDiffNs.get(storeName).update(System.nanoTime() - dirDiffStartTime);
+
+        DirDiff.Stats stats = DirDiff.getStats(dirDiff);
+        updateStoreDiffMetrics(storeName, stats);
+        metrics.filesToUpload.getValue().addAndGet(stats.filesAdded);
```
[the remark this answers is not in the document]
Comment (prateekm) [reply samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreBackupManager.java:205]: WDYM by moved to stats class?

File: samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreBackupManager.java [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+   */
+  @Override
+  public CompletableFuture<Void> cleanUp(CheckpointId checkpointId, Map<String, String> storeSCMs) {
+    long startTime = System.nanoTime();
+    List<CompletionStage<Void>> removeTTLFutures = new ArrayList<>();
+    List<CompletionStage<Void>> cleanupRemoteSnapshotFutures = new ArrayList<>();
+    List<CompletionStage<Void>> removePrevRemoteSnapshotFutures = new ArrayList<>();
+
+    // SCM, in case of blob store backup and restore, is just the blob id of SnapshotIndex representing the remote snapshot
+    storeSCMs.forEach((storeName, snapshotIndexBlobId) -> {
```
[the remark this answers is not in the document]
Comment (prateekm) [reply samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreBackupManager.java:283]: Can you clarify what you changed, if anything, and why? Is this an issue or not?

Use storesToBackup map instead of creating again?

File: samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreBackupManager.java [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+      }
+    });
+
+    return FutureUtil.allOf(removeTTLFutures, cleanupRemoteSnapshotFutures, removePrevRemoteSnapshotFutures)
+        .whenComplete((res, ex) -> metrics.cleanupNs.update(System.nanoTime() - startTime));
+  }
+
+  @Override
+  public void close() {
+    blobStoreUtil.closeBlobStoreManager();
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreBackupManager.java:333]: Why is BlobStoreUtil proxying the close call to the Manager?

File: samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreRestoreManager.java [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+  /**
+   * Determines if the store needs to be restored from remote snapshot based on local and remote state.
+   */
+  @VisibleForTesting
+  static boolean shouldRestore(String taskName, String storeName, DirIndex dirIndex,
+      Path storeCheckpointDir, StorageConfig storageConfig, BlobStoreUtil blobStoreUtil) {
+    // if a store checkpoint directory exists for the last successful task checkpoint, try to use it.
+    boolean restoreStore;
+    if (Files.exists(storeCheckpointDir)) {
+      if (storageConfig.getCleanLoggedStoreDirsOnStart(storeName)) {
```
[the remark this answers is not in the document]
Comment (prateekm) [reply samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreRestoreManager.java:288]: Accessors for boolean values should be named `X / isX / shouldX / doX` etc., not getX etc.

E.g., isSkyBlue makes more sense than getIsSkyBlue, shouldRun makes more sense than getShouldRun etc.

File: samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreStateBackendFactory.java (hunk 4 above)
Comment (prateekm) [reply samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreStateBackendFactory.java:68]: +1 for keeping them separate, since some of the metrics have similar names and functionality, just in opposite directions (upload vs download). Agree that it's not that nice for BlobStoreUtil, but I prefer this tradeoff too.

File: samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala [hunk 12]
```diff
@@ -500,8 +519,9 @@ object SamzaContainer extends Logging {
     val loggedStorageBaseDir = getLoggedStorageBaseDir(jobConfig, defaultStoreBaseDir)
     info("Got base directory for logged data stores: %s" format loggedStorageBaseDir)
 
+    // TODO dchen should we enforce restore factories to be subset of backup factories?
```
Comment (prateekm) [inline samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala:522]: TODO HIGH dchen. @dxichen let's fix this in the "config scope and precedence" follow up for async commit.

File: samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala [hunk 13]
```diff
@@ -719,7 +745,11 @@ class SamzaContainer(
 
   private val jobConfig = new JobConfig(config)
   private val taskConfig = new TaskConfig(config)
-  val shutdownMs: Long = taskConfig.getShutdownMs
+
+  // Linkedin-specific shutdownMs due to SAMZA-2198; switch back to TaskConfig.getShutdownMs once that is fixed
+  val shutdownMs: Long = taskConfig.getLong(TaskConfig.TASK_SHUTDOWN_MS, 5000)
```
Comment (prateekm) [inline samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala:750]: Does this need to be here?

File: samza-core/src/main/java/org/apache/samza/config/BlobStoreConfig.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+      "org.apache.samza.storage.blobstore.BlobStoreStateBackendFactory";
+  public static final String BLOB_STORE_MANAGER_FACTORY = "blob.store.manager.factory";
+  public static final String BLOB_STORE_STATE_BACKEND_ADMIN_FACTORY = "blob.store.state.backend.admin.factory";
+
+  public static final String DEFAULT_STORE_STATE_BACKEND_FACTORY =
+      "org.apache.samza.storage.KafkaChangelogStateBackendFactory";
+  public static final List<String> DEFAULT_STORE_STATE_BACKEND_BACKUP_FACTORIES = ImmutableList.of(
+      DEFAULT_STORE_STATE_BACKEND_FACTORY);
+  public static final String STORE_STATE_BACKEND_BACKUP_FACTORIES = STORE_PREFIX + "%s.state.backend.backup.factories";
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/config/BlobStoreConfig.java:26]: BLOCKER: These are not blob store manager specific, these should be in storage config.

Only the getBlobStoreManagerFactory, getBlobStoreAdminFactory should be here.

**Review by prateekm — CHANGES_REQUESTED** [review]

File: samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreBackupManager.java [hunk 15]
```diff
@@ -0,0 +1,349 @@
...
+
+        // upload the new snapshot index to the blob store and get its blob id
+        CompletionStage<String> snapshotIndexBlobIdFuture =
+            snapshotIndexFuture
+                .thenComposeAsync(si -> {
+                  LOG.trace("Uploading Snapshot index for task: {} store: {}", taskName, storeName);
+                  return blobStoreUtil.putSnapshotIndex(si);
+                }, executor);
+
+        // update the map of storeName to previous snapshot index storeToSCMAndSnapshotIndexPairFutures which is temporary
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreBackupManager.java:233]: Run on sentence is hard to parse, can you simplify? Don't need to mention variable names in comment, it's obvious from context. Just say what is happening here and why. 

There is actually no update happening here at all, is this comment misplaced (should be on 244)?

Maybe we can delete the comment entirely?

File: samza-core/src/main/java/org/apache/samza/storage/blobstore/metrics/BlobStoreRestoreManagerMetrics.java [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+import org.apache.samza.metrics.Gauge;
+import org.apache.samza.metrics.MetricsRegistry;
+import org.apache.samza.metrics.Timer;
+
+
+public class BlobStoreRestoreManagerMetrics {
+  private static final String GROUP = BlobStoreRestoreManagerMetrics.class.getName();
+  private final MetricsRegistry metricsRegistry;
+
+  // ToDo per-task throughput
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/storage/blobstore/metrics/BlobStoreRestoreManagerMetrics.java:17]: Update TODO format (here and everywhere else).

File: samza-core/src/main/java/org/apache/samza/storage/blobstore/util/DirDiffUtil.java [hunk 17]
```diff
@@ -0,0 +1,404 @@
...
+import java.util.HashMap;
+import java.util.zip.CRC32;
+import java.util.zip.CheckedInputStream;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.samza.SamzaException;
+import org.apache.samza.storage.blobstore.diff.DirDiff;
+import org.apache.samza.storage.blobstore.index.DirIndex;
+import org.apache.samza.storage.blobstore.index.FileIndex;
+
+import java.io.File;
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/storage/blobstore/util/DirDiffUtil.java:38]: Sort imports (you can do this for entire directory at once in IntelliJ (Right click - > Optimize imports).

File: samza-core/src/main/java/org/apache/samza/config/StorageConfig.java [hunk 18]
```diff
@@ -66,6 +66,9 @@
   public static final String CHANGELOG_MIN_COMPACTION_LAG_MS = STORE_PREFIX + "%s.changelog." + MIN_COMPACTION_LAG_MS;
   public static final long DEFAULT_CHANGELOG_MIN_COMPACTION_LAG_MS = TimeUnit.HOURS.toMillis(4);
 
+  public static final String BLOB_STORE_BACKEND_ADMIN_FACTORY = "blob.store.backend.admin.factory";
```
[the remark this answers is not in the document]
Comment (prateekm) [reply samza-core/src/main/java/org/apache/samza/config/StorageConfig.java]: @shekhars-li Let's rename the configs / fields / accessors for consistency too. E.g. maybe:
```stores.%s.backup.factories 
stores.%s.restore.factories 
blob.store.admin.factory 
blob.store.manager.factory
getStoresWithBackupFactory()
getStoresWithRestoreFactory()
``` etc.

File: samza-api/src/main/java/org/apache/samza/storage/blobstore/exceptions/DeletedException.java [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.storage.blobstore.exceptions;
+
+public class DeletedException extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (dxichen) [reply samza-api/src/main/java/org/apache/samza/storage/blobstore/exceptions/DeletedException.java:22]: P2: We would need to document in both BlobStoreManager as well as in this class exactly the expectation of how this class would be used in the BlobStoreManager so that implementers of that and this interface could have that background

File: samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreBackupManager.java (hunk 10 above)
Comment (shekhars-li) [reply samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreBackupManager.java]: Fixed that to call init/close directly.

**Comment by Sanil15** [comment]
Some minor comments

**Comment by prateekm** [comment]
@Sanil15 FYI, @shekhars-li will address rest of the feedback in a follow up PR.
